app/validators/article_validator.rb

```diff
@@ -0,0 +1,10 @@
+class ArticleValidator < ActiveModel::Validator
+  def validate(record)
+    unless Article.find_by(short_path: record.short_path, id: record.id)
```

Comment:
Why `find_by` `id` and `short_path`? The next find also finds by `short_path` and you already know the article with that `id`, it's `record`. Unless there is some nuance I'm missing here I'm guessing you could get away with just `validates :short_path, uniqueness: true` on `article.rb` and not have this custom validator. You could also add a similar validator on `Redirect` and handle the error `create_redirect`if it fails to save.